src/object/custom_particle_system.hpp

```diff
@@ -0,0 +1,408 @@
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#ifndef HEADER_SUPERTUX_OBJECT_CUSTOM_PARTICLE_SYSTEM_HPP
+#define HEADER_SUPERTUX_OBJECT_CUSTOM_PARTICLE_SYSTEM_HPP
+
+#include "math/easing.hpp"
+#include "math/vector.hpp"
+#include "object/particlesystem_interactive.hpp"
+#include "object/particle_zone.hpp"
+#include "scripting/custom_particles.hpp"
+#include "video/surface.hpp"
+#include "video/surface_ptr.hpp"
+
+class CustomParticleSystem :
+  public ParticleSystem_Interactive,
+  public ExposedObject<CustomParticleSystem, scripting::CustomParticles>
+{
+  friend class ParticleEditor;
+public:
+  CustomParticleSystem();
+  CustomParticleSystem(const ReaderMapping& reader);
+  virtual ~CustomParticleSystem();
+
+  virtual void draw(DrawingContext& context) override;
+
+  void reinit_textures();
+  virtual void update(float dt_sec) override;
+
+  virtual std::string get_class() const override { return "particles-custom"; }
+  virtual std::string get_display_name() const override { return _("Custom Particles"); }
+  virtual void save(Writer& writer) override;
+  virtual ObjectSettings get_settings() override;
+
+  virtual const std::string get_icon_path() const override {
+    return "images/engine/editor/sparkle.png";
+  }
+
+  //void fade_amount(int new_amount, float fade_time);
+protected:
+  virtual int collision(Particle* particle, const Vector& movement) override;
+
+private:
+
+  // Local
+  void add_particle(float lifetime, float x, float y);
+  void spawn_particles(float lifetime);
+
+  std::vector<ParticleZone::ZoneDetails> get_zones();
+
+  float get_abs_x();
+  float get_abs_y();
+  
+  float texture_sum_odds;
+  float time_last_remaining;
+
+public:
+  // Scripting
+  void clear() { custom_particles.clear(); }
+
+private:
+  enum class RotationMode {
+    Fixed,
+    Facing,
+    Wiggling
+  };
+
+  enum class FadeMode {
+    None,
+    Fade,
+    Shrink
+  };
+
+  enum class CollisionMode {
+    Ignore,
+    Stick,
+    BounceHeavy,
+    BounceLight,
+    Destroy
+  };
+
+  enum class OffscreenMode {
+    Never,
+    OnlyOnExit,
+    Always
+  };
+
+  class SpriteProperties final
+  {
+  public:
+    float likeliness;
+    Color color;
+    SurfacePtr texture;
+    Vector scale;
+
+    SpriteProperties() :
+      likeliness(1.f),
+      color(1.f, 1.f, 1.f, 1.f),
+      texture(Surface::from_file("images/engine/editor/sparkle.png")),
+      scale(1.f, 1.f)
+    {
+    }
+
+    SpriteProperties(SurfacePtr surface) :
+      likeliness(1.f),
+      color(1.f, 1.f, 1.f, 1.f),
+      texture(surface),
+      scale(1.f, 1.f)
+    {
+    }
+
+    SpriteProperties(SpriteProperties& sp, float alpha) :
```

Comment:
Can this be `const SpriteProperties& sp`?

Reply:
Yes :+1: